### ISSUE TITLE
3.0 - When the primary key type of an entity is UUID, a new value is always generated even if a value is set before saving

### ISSUE THREAD

**Issue by akinwale** [cakephp/cakephp#7225]
Assuming I have the following table definition.
```sql
CREATE TABLE `Items`
(
    `Guid` CHAR(36) NOT NULL PRIMARY KEY,
    `Name` VARCHAR(100)
);
```

With my ItemsTable class defined and the primary key set to the Guid field, every time I try to call the save method on a new entity, a new value is generated which overwrites the data value.

For example:
```php
$this->loadModel('Items');
$data = ['Guid' => Text::uuid(), 'Name' => 'First Item'];
$item = $this->Items->newEntity($data);
$this->Items->save($item);
// At this point, $item->Guid is not the same value as $data['Guid']
```

I was able to trace this to `Cake\ORM\Table#1489`
```php
$id = (array)$this->_newId($primary) + $keys;
```

My workaround was to obtain the field value from the `$data` array if set. However, this obviously does not support composite keys.
```php
$id = null;
if (count($primary) == 1 && isset($data[$primary[0]])) {
    $id = (array)$data[$primary[0]] + $keys;
} else {
    $id = (array)$this->_newId($primary) + $keys;
}
```

**Comment by robertpustulka** [comment]
CakePHP can handle uuid PKs by itself, you don't have to generate it manually.

**Comment by AD7six** [comment]
@robertpustulka should be possible to insert with a specific id though

**Comment by robertpustulka** [comment]
@AD7six Would it work if you did this: `$entity->isNew(false)`?

**Comment by markstory** [comment]
Inserting with a specific ID is not possible for autoincrement keys though. Both Postgres and SQLserver disallow setting specific values.

Changing the code as described above might cause issues with autoincrements, so I'm not sure its quite as simple as it initially seems, but I might be wrong.

**Comment by akinwale** [comment]
Indeed. I would imagine adding checks for the primary key metadata to verify if it's an identity column (SQL Server) or a serial (PostgreSQL). My code is simply a workaround for the project I'm currently working on until there's a proper fix.

**Comment by ndm2** [comment]
@akinwale For something that is less of a workaround, you could instead override `Table::_newId()` to return `null`, this will cause the possible primary key set in the data to be used. All non-UUID database types currently do that, they return `null` for new IDs.

**Comment by robertscherer** [comment]
Wasn't this fixed in https://github.com/cakephp/cakephp/issues/6185 ?

**Comment by markstory** [comment]
@robertscherer I would have thought so too. I'll take another look and make sure the tests added in #6815 are still included and cover this case.

**Comment by markstory** [comment]
So saving with a specific UUID is possible. However, the entity has its `id` overwritten making it look like the original id was not used.

```
>>> $id = \Cake\Utility\Text::uuid();
=> "286a3a63-e5d6-4fe8-89de-bfb7767e6be4"
>>> $items = \Cake\ORM\TableRegistry::get('Items');
=> Cake\ORM\Table {#214
     +"registryAlias": "Items",
     +"table": "items",
     +"alias": "Items",
     +"entityClass": "\Cake\ORM\Entity",
     +"associations": [],
     +"behaviors": [],
     +"defaultConnection": "default",
     +"connectionName": "default",
   }
>>> $item = new \Cake\ORM\Entity(['id' => $id, 'name' => '123abc']);
=> Cake\ORM\Entity {#223
     +"id": "286a3a63-e5d6-4fe8-89de-bfb7767e6be4",
     +"name": "123abc",
     +"[new]": true,
     +"[accessible]": [
       "*" => true,
     ],
     +"[dirty]": [
       "id" => true,
       "name" => true,
     ],
     +"[original]": [],
     +"[virtual]": [],
     +"[errors]": [],
     +"[repository]": null,
   }
>>> $items->save($item);
=> Cake\ORM\Entity {#223
     +"id": "103f6d73-1917-4993-9cd4-c95e6b8aeb46",
     +"name": "123abc",
     +"[new]": false,
     +"[accessible]": [
       "*" => true,
     ],
     +"[dirty]": [],
     +"[original]": [
       "id" => "286a3a63-e5d6-4fe8-89de-bfb7767e6be4",
     ],
     +"[virtual]": [],
     +"[errors]": [],
     +"[repository]": "Items",
   }
>>> $items->find()->where(['id' => $id])->count();
=> 1
```

As you can see the `id` was persisted, but the entity has the incorrect id attribute post save().  I'll get this fixed up.

**Comment by markstory** [comment]
Fixed with #7322
